### ISSUE TITLE
Bug in `with` parameter of `cross` transform

### ISSUE THREAD

**Issue by Jerrythafast** [vega/vega#271]
Dear Vega developers,

Today I stumbled into an issue with the `with` parameter of the `cross` transform. The [Wiki](https://github.com/trifacta/vega/wiki/Data-Transforms#-cross) says:

> **with** - [String] The name of the secondary data set to cross with the primary data. If unspecified, the primary data is crossed with itself.

This, however, does not seem to be the case. It appears to me that the value of the `with` parameter is treated like a data set directly.

```json
"data": [
  {
    "name": "testX",
    "values": [1,2,3,4,5]
  },
  {
    "name": "testY",
    "values": [1,2,3,4,5],
    "transform": [{"type": "cross", "with": "testX"}]
  }
]
```

This results in the following (note the values of `testY.b`):
```json
{
  "testX": [
    {"data": 1, "index": 0},
    {"data": 2, "index": 1},
    {"data": 3, "index": 2},
    {"data": 4, "index": 3},
    {"data": 5, "index": 4}
  ],
  "testY": [
    {"a": {"data": 1, "index": 0}, "b": "t"},
    {"a": {"data": 1, "index": 0}, "b": "e"},
    {"a": {"data": 1, "index": 0}, "b": "s"},
    {"a": {"data": 1, "index": 0}, "b": "t"},
    {"a": {"data": 1, "index": 0}, "b": "X"},
    {"a": {"data": 2, "index": 1}, "b": "t"},
    {"a": {"data": 2, "index": 1}, "b": "e"},
    {"a": {"data": 2, "index": 1}, "b": "s"},
    {"a": {"data": 2, "index": 1}, "b": "t"},
    {"a": {"data": 2, "index": 1}, "b": "X"},
    {"a": {"data": 3, "index": 2}, "b": "t"},
    {"a": {"data": 3, "index": 2}, "b": "e"},
    {"a": {"data": 3, "index": 2}, "b": "s"},
    {"a": {"data": 3, "index": 2}, "b": "t"},
    {"a": {"data": 3, "index": 2}, "b": "X"},
    {"a": {"data": 4, "index": 3}, "b": "t"},
    {"a": {"data": 4, "index": 3}, "b": "e"},
    {"a": {"data": 4, "index": 3}, "b": "s"},
    {"a": {"data": 4, "index": 3}, "b": "t"},
    {"a": {"data": 4, "index": 3}, "b": "X"},
    {"a": {"data": 5, "index": 4}, "b": "t"},
    {"a": {"data": 5, "index": 4}, "b": "e"},
    {"a": {"data": 5, "index": 4}, "b": "s"},
    {"a": {"data": 5, "index": 4}, "b": "t"},
    {"a": {"data": 5, "index": 4}, "b": "X"}
  ]
}
```

Now, I must say I actually found this behaviour useful. In fact I was able to plot some polynomials in Vega using a surprisingly clean syntax thanks to this bug! The following yields a nice data set where each datum has values `x`, `y`, and `coefs`. You can easily `facet` on `coefs` to plot `x` vs `y`.

```json
"data": [
  {
    "name": "polynomials",
    "values": [
      {"a": 3, "b": 2, "c": 9},
      {"a": 2, "b": 2, "c": 1},
      {"a": 5, "b": 1, "c": 2},
      {"a": 1, "b": 7, "c": 6},
      {"a": 1, "b": 6, "c": 4},
      {"a": 3, "b": 3, "c": 3}
    ],
    "transform": [
      {
        "type": "cross",
        "with": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
        "output": {"left": "coefs", "right": "x"}
      },
      {
        "type": "formula",
        "field": "y",
        "expr": "d.coefs.data.a*d.x*d.x + d.coefs.data.b*d.x + d.coefs.data.c"
      }
    ]
  }
]
```

The 'official' (but broken) way to do the same thing is:

```json
"data": [
  {
    "name": "coefficients",
    "values": [
      {"a": 3, "b": 2, "c": 9},
      {"a": 2, "b": 2, "c": 1},
      {"a": 5, "b": 1, "c": 2},
      {"a": 1, "b": 7, "c": 6},
      {"a": 1, "b": 6, "c": 4},
      {"a": 3, "b": 3, "c": 3}
    ]
  },
  {
    "name": "polynomials",
    "values": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
    "transform": [
      {
        "type": "cross",
        "with": "coefficients",
        "output": {"left": "x", "right": "coefs"}
      },
      {
        "type": "formula",
        "field": "y",
        "expr": "d.coefs.data.a*d.x*d.x + d.coefs.data.b*d.x + d.coefs.data.c"
      }
    ]
  }
]
```

So perhaps you add this possibility through an official parameter as well :wink:

Yours,
Jerry

**Comment by Jerrythafast** [comment]
Oh, I just realized that the cross transform also assumes that the `with` dataset is assumed to be of the same size as the current dataset. So the above example actually only computes the polynomials up to x=5 because there are 6 polynomials in there. If the `with` dataset is too short you get `x=undefined` in the output, and the ` formula` then results in a `NaN`.

**Comment by arvind** [comment]
Thanks for the bug report. This should be fixed in Vega 2.0, with unit tests to verify :) Please re-open the issue if you still encounter it!
